File: legacy-python/cursor_orch/api/repo_store.py
from __future__ import annotations

import base64
import logging
import random
import time

import requests

logger = logging.getLogger(__name__)

BASE_URL = "https://api.github.com"

INITIAL_DELAY = 5.0
BACKOFF_MULTIPLIER = 2.0
MAX_DELAY = 60.0
MAX_RETRIES_429 = 5
MAX_RETRIES_TRANSIENT = 3
MAX_RETRIES_409 = 3
MAX_RETRIES_NETWORK = 3
JITTER_FACTOR = 0.2
TRANSIENT_CODES = (502, 503, 504)


class RepoStoreError(Exception):
    def __init__(self, message: str, status_code: int = 0):
        super().__init__(message)
        self.message = message
        self.status_code = status_code


class RepoStoreNotFound(RepoStoreError):
    def __init__(self, path: str = ""):
        super().__init__(f"Not found: {path}", 404)


class RateLimitError(RepoStoreError):
    def __init__(self, message: str = "Rate limit exceeded", retry_after: float = 0):
        super().__init__(message, 429)
        self.retry_after = retry_after


def _encode_content(content: str) -> str:
    return base64.b64encode(content.encode("utf-8")).decode("ascii")


def _decode_content(encoded: str) -> str:
    return base64.b64decode(encoded).decode("utf-8")


def _compute_delay(attempt: int) -> float:
    delay = min(INITIAL_DELAY * (BACKOFF_MULTIPLIER ** attempt), MAX_DELAY)
    jitter = delay * JITTER_FACTOR
    return delay + random.uniform(-jitter, jitter)


def _compute_429_delay(attempt: int, resp: requests.Response) -> float:
    delay = _compute_delay(attempt)
    retry_after = resp.headers.get("Retry-After")
    if retry_after is not None:
        delay = max(delay, float(retry_after))
    return delay


class RepoStoreClient:
    def __init__(self, token: str, owner: str, repo: str):
        self._session = requests.Session()
        self._session.headers.update({
            "Authorization": f"token {token}",
            "Accept": "application/vnd.github+json",
        })
        self._owner = owner
        self._repo = repo
        self._rate_limit_remaining: int | None = None
        self._rate_limit_limit: int | None = None
        self._rate_limit_reset: int | None = None

    @property
    def rate_limit_remaining(self) -> int | None:
        return self._rate_limit_remaining

    def _track_rate_limit(self, resp: requests.Response) -> None:
        remaining = resp.headers.get("X-RateLimit-Remaining")
        if remaining is not None:
            self._rate_limit_remaining = int(remaining)
        limit = resp.headers.get("X-RateLimit-Limit")
        if limit is not None:
            self._rate_limit_limit = int(limit)
        reset = resp.headers.get("X-RateLimit-Reset")
        if reset is not None:
            self._rate_limit_reset = int(reset)

    def _request(self, method: str, url: str, **kwargs: object) -> requests.Response:
        retries_429 = 0
        retries_transient = 0
        retries_network = 0

        while True:
            try:
                resp = self._session.request(method, url, **kwargs)
            except requests.exceptions.RequestException as exc:
                if retries_network < MAX_RETRIES_NETWORK:
                    delay = _compute_delay(retries_network)
                    retries_network += 1
                    logger.warning(
                        "GitHub API network error (%s). Retry %s/%s in %.1fs",
                        type(exc).__name__,
                        retries_network,
                        MAX_RETRIES_NETWORK,
                        delay,
                    )
                    time.sleep(delay)
                    continue
                raise RepoStoreError(
                    f"GitHub API network error after max retries: {type(exc).__name__}: {exc}"
                ) from exc
            self._track_rate_limit(resp)

            if resp.status_code == 429 and retries_429 < MAX_RETRIES_429:
                delay = _compute_429_delay(retries_429, resp)
                retries_429 += 1
                logger.warning("Rate limited (429). Retry %s/%s in %.1fs", retries_429, MAX_RETRIES_429, delay)
                time.sleep(delay)
                continue

            if resp.status_code == 429:
                raise RateLimitError("GitHub API rate limit exceeded after max retries")

            if resp.status_code in TRANSIENT_CODES and retries_transient < MAX_RETRIES_TRANSIENT:
                delay = _compute_delay(retries_transient)
                retries_transient += 1
                logger.warning("Transient error (%s). Retry %s/%s in %.1fs", resp.status_code, retries_transient, MAX_RETRIES_TRANSIENT, delay)
                time.sleep(delay)
                continue

            if resp.status_code in TRANSIENT_CODES:
                raise RepoStoreError(f"Transient error {resp.status_code} after max retries", resp.status_code)

            if resp.status_code == 404:
                raise RepoStoreNotFound()

            if resp.status_code >= 400:
                raise RepoStoreError(f"GitHub API error: {resp.status_code} {resp.text[:500]}", resp.status_code)

            return resp

    def _get_default_branch_sha(self) -> str:
        url = f"{BASE_URL}/repos/{self._owner}/{self._repo}"
        resp = self._request("GET", url)
        default_branch: str = resp.json().get("default_branch", "main")
        ref_url = f"{BASE_URL}/repos/{self._owner}/{self._repo}/git/ref/heads/{default_branch}"
        ref_resp = self._request("GET", ref_url)
        sha: str = ref_resp.json()["object"]["sha"]
        return sha

    def _get_file_sha(self, run_id: str, filename: str) -> str | None:
        url = f"{BASE_URL}/repos/{self._owner}/{self._repo}/contents/{filename}"
        try:
            resp = self._request("GET", url, params={"ref": f"run/{run_id}"})
        except RepoStoreNotFound:
            return None
        data = resp.json()
        sha: str | None = data.get("sha")
        return sha

    def create_run(self, run_id: str) -> None:
        sha = self._get_default_branch_sha()
        url = f"{BASE_URL}/repos/{self._owner}/{self._repo}/git/refs"
        payload = {"ref": f"refs/heads/run/{run_id}", "sha": sha}
        try:
            self._request("POST", url, json=payload)
        except RepoStoreError as exc:
            if exc.status_code == 422:
                logger.info("Branch run/%s already exists, skipping", run_id)
                return
            raise

    def read_file(self, run_id: str, filename: str) -> str:
        url = f"{BASE_URL}/repos/{self._owner}/{self._repo}/contents/{filename}"
        try:
            resp = self._request("GET", url, params={"ref": f"run/{run_id}"})
        except RepoStoreNotFound:
            return ""
        data = resp.json()
        encoded: str = data.get("content", "")
        return _decode_content(encoded.replace("\n", ""))

    def write_file(self, run_id: str, filename: str, content: str) -> None:
        url = f"{BASE_URL}/repos/{self._owner}/{self._repo}/contents/{filename}"
        branch = f"run/{run_id}"
        encoded = _encode_content(content)

        for attempt in range(MAX_RETRIES_409 + 1):
            sha = self._get_file_sha(run_id, filename)
            payload: dict[str, object] = {
                "message": f"update {filename}",
                "content": encoded,
                "branch": branch,
            }
            if sha is not None:
                payload["sha"] = sha

            try:
                self._request("PUT", url, json=payload)
                return
            except RepoStoreError as exc:
                if exc.status_code in (409, 422) and attempt < MAX_RETRIES_409:
                    delay = _compute_delay(attempt)
                    logger.warning(
                        "Write conflict (%s) for %s. Retry %s/%s in %.1fs",
                        exc.status_code,
                        filename,
                        attempt + 1,
                        MAX_RETRIES_409,
                        delay,
                    )
                    time.sleep(delay)
                    continue
                raise

    def delete_file(self, run_id: str, filename: str) -> None:
        sha = self._get_file_sha(run_id, filename)
        if sha is None:
            return
        url = f"{BASE_URL}/repos/{self._owner}/{self._repo}/contents/{filename}"
        payload: dict[str, object] = {
            "message": f"delete {filename}",
            "sha": sha,
            "branch": f"run/{run_id}",
        }
        try:
            self._request("DELETE", url, json=payload)
        except RepoStoreNotFound:
            return

    def list_run_files(self, run_id: str) -> list[str]:
        url = f"{BASE_URL}/repos/{self._owner}/{self._repo}/contents/"
        try:
            resp = self._request("GET", url, params={"ref": f"run/{run_id}"})
        except RepoStoreNotFound:
            return []
        items: list[dict] = resp.json()
        return [item["name"] for item in items if isinstance(item, dict)]

    def _get_branch_sha(self, run_id: str) -> str:
        url = f"{BASE_URL}/repos/{self._owner}/{self._repo}/git/ref/heads/run/{run_id}"
        resp = self._request("GET", url)
        sha: str = resp.json()["object"]["sha"]
        return sha

    def _create_blob(self, content: str) -> str:
        url = f"{BASE_URL}/repos/{self._owner}/{self._repo}/git/blobs"
        resp = self._request("POST", url, json={"content": content, "encoding": "utf-8"})
        sha: str = resp.json()["sha"]
        return sha

    def _create_tree(self, file_blobs: dict[str, str]) -> str:
        tree = [
            {"path": path, "mode": "100644", "type": "blob", "sha": blob_sha}
            for path, blob_sha in file_blobs.items()
        ]
        url = f"{BASE_URL}/repos/{self._owner}/{self._repo}/git/trees"
        resp = self._request("POST", url, json={"tree": tree})
        sha: str = resp.json()["sha"]
        return sha

    def _create_commit(self, message: str, tree_sha: str, parent_sha: str) -> str:
        url = f"{BASE_URL}/repos/{self._owner}/{self._repo}/git/commits"
        payload = {"message": message, "tree": tree_sha, "parents": [parent_sha]}
        resp = self._request("POST", url, json=payload)
        sha: str = resp.json()["sha"]
        return sha

    def _update_branch_ref(self, run_id: str, commit_sha: str) -> None:
        url = f"{BASE_URL}/repos/{self._owner}/{self._repo}/git/refs/heads/run/{run_id}"
        self._request("PATCH", url, json={"sha": commit_sha})

    def list_run_branches(self) -> list[str]:
        url = f"{BASE_URL}/repos/{self._owner}/{self._repo}/git/refs/heads/run/"
        try:
            resp = self._request("GET", url)
        except RepoStoreNotFound:
            return []
        items = resp.json()
        if not isinstance(items, list):
            return []
        return [
            item["ref"].removeprefix("refs/heads/")
            for item in items
            if isinstance(item, dict) and "ref" in item
        ]

    def delete_run_branch(self, run_id: str) -> None:
        url = f"{BASE_URL}/repos/{self._owner}/{self._repo}/git/refs/heads/run/{run_id}"
        try:
            self._request("DELETE", url)
        except RepoStoreNotFound:
            pass

    def batch_write_files(self, run_id: str, files: dict[str, str], message: str = "batch update") -> None:
        for attempt in range(MAX_RETRIES_409 + 1):
            parent_sha = self._get_branch_sha(run_id)
            file_blobs = {path: self._create_blob(content) for path, content in files.items()}
            tree_sha = self._create_tree(file_blobs)
            commit_sha = self._create_commit(message, tree_sha, parent_sha)
            try:
                self._update_branch_ref(run_id, commit_sha)
                return
            except RepoStoreError as exc:
                if exc.status_code == 422 and attempt < MAX_RETRIES_409:
                    delay = _compute_delay(attempt)
                    logger.warning(
                        "Ref update conflict (422). Retry %s/%s in %.1fs",
                        attempt + 1,
                        MAX_RETRIES_409,
                        delay,
                    )
                    time.sleep(delay)
                    continue
                raise

File: src/api/repo-store.ts
import { setTimeout as delay } from "node:timers/promises";

const BASE_URL = "https://api.github.com";

const INITIAL_DELAY = 5.0;
const BACKOFF_MULTIPLIER = 2.0;
const MAX_DELAY = 60.0;
const MAX_RETRIES_429 = 5;
const MAX_RETRIES_TRANSIENT = 3;
const MAX_RETRIES_409 = 3;
const MAX_RETRIES_NETWORK = 3;
const JITTER_FACTOR = 0.2;
const TRANSIENT_CODES = new Set([502, 503, 504]);

export class RepoStoreError extends Error {
  statusCode: number;
  constructor(message: string, statusCode = 0) {
    super(message);
    this.name = "RepoStoreError";
    this.statusCode = statusCode;
  }
}

export class RepoStoreNotFound extends RepoStoreError {
  constructor(path = "") {
    super(`Not found: ${path}`, 404);
    this.name = "RepoStoreNotFound";
  }
}

export class RateLimitError extends RepoStoreError {
  retryAfter: number;
  constructor(message = "Rate limit exceeded", retryAfter = 0) {
    super(message, 429);
    this.name = "RateLimitError";
    this.retryAfter = retryAfter;
  }
}

function encodeContent(content: string): string {
  return Buffer.from(content, "utf8").toString("base64");
}

function decodeContent(encoded: string): string {
  return Buffer.from(encoded.replace(/\n/g, ""), "base64").toString("utf8");
}

function computeDelay(attempt: number): number {
  const d = Math.min(INITIAL_DELAY * BACKOFF_MULTIPLIER ** attempt, MAX_DELAY);
  const jitter = d * JITTER_FACTOR;
  return d + (Math.random() * 2 - 1) * jitter;
}

function compute429Delay(attempt: number, retryAfterHeader: string | null): number {
  let d = computeDelay(attempt);
  if (retryAfterHeader !== null) {
    const v = parseFloat(retryAfterHeader);
    if (!Number.isNaN(v)) {
      d = Math.max(d, v);
    }
  }
  return d;
}

export class RepoStoreClient {
  private readonly token: string;
  private readonly owner: string;
  private readonly repo: string;
  private _rateLimitRemaining: number | null = null;
  private _rateLimitLimit: number | null = null;

  constructor(token: string, owner: string, repo: string) {
    this.token = token;
    this.owner = owner;
    this.repo = repo;
  }

  get rateLimitRemaining(): number | null {
    return this._rateLimitRemaining;
  }

  get rateLimitLimit(): number | null {
    return this._rateLimitLimit;
  }

  private trackRateLimit(resp: Response): void {
    const remaining = resp.headers.get("X-RateLimit-Remaining");
    if (remaining !== null) {
      this._rateLimitRemaining = parseInt(remaining, 10);
    }
    const limit = resp.headers.get("X-RateLimit-Limit");
    if (limit !== null) {
      this._rateLimitLimit = parseInt(limit, 10);
    }
  }

  private async request(method: string, url: string, body?: unknown): Promise<Response> {
    let retries429 = 0;
    let retriesTransient = 0;
    let retriesNetwork = 0;
    const headers: Record<string, string> = {
      Authorization: `token ${this.token}`,
      Accept: "application/vnd.github+json",
    };
    if (body !== undefined) {
      headers["Content-Type"] = "application/json";
    }
    while (true) {
      try {
        const resp = await fetch(url, {
          method,
          headers,
          body: body !== undefined ? JSON.stringify(body) : undefined,
        });
        this.trackRateLimit(resp);
        if (resp.status === 429 && retries429 < MAX_RETRIES_429) {
          const d = compute429Delay(retries429, resp.headers.get("Retry-After"));
          retries429 += 1;
          await delay(Math.ceil(d * 1000));
          continue;
        }
        if (resp.status === 429) {
          throw new RateLimitError("GitHub API rate limit exceeded after max retries");
        }
        if (TRANSIENT_CODES.has(resp.status) && retriesTransient < MAX_RETRIES_TRANSIENT) {
          const d = computeDelay(retriesTransient);
          retriesTransient += 1;
          await delay(Math.ceil(d * 1000));
          continue;
        }
        if (TRANSIENT_CODES.has(resp.status)) {
          throw new RepoStoreError(`Transient error ${resp.status} after max retries`, resp.status);
        }
        if (resp.status === 404) {
          throw new RepoStoreNotFound();
        }
        if (resp.status >= 400) {
          const text = await resp.text();
          throw new RepoStoreError(`GitHub API error: ${resp.status} ${text.slice(0, 500)}`, resp.status);
        }
        return resp;
      } catch (e) {
        if (e instanceof RepoStoreError) {
          throw e;
        }
        if (retriesNetwork < MAX_RETRIES_NETWORK) {
          const d = computeDelay(retriesNetwork);
          retriesNetwork += 1;
          await delay(Math.ceil(d * 1000));
          continue;
        }
        throw new RepoStoreError(`GitHub API network error after max retries: ${e}`);
      }
    }
  }

  private async getDefaultBranchSha(): Promise<string> {
    const url = `${BASE_URL}/repos/${this.owner}/${this.repo}`;
    const resp = await this.request("GET", url);
    const data = (await resp.json()) as { default_branch?: string };
    const defaultBranch = data.default_branch ?? "main";
    const refUrl = `${BASE_URL}/repos/${this.owner}/${this.repo}/git/ref/heads/${defaultBranch}`;
    const refResp = await this.request("GET", refUrl);
    const refData = (await refResp.json()) as { object?: { sha?: string } };
    return refData.object?.sha ?? "";
  }

  private async getFileSha(runId: string, filename: string): Promise<string | null> {
    const url = `${BASE_URL}/repos/${this.owner}/${this.repo}/contents/${filename}?ref=${encodeURIComponent(`run/${runId}`)}`;
    try {
      const resp = await this.request("GET", url);
      const data = (await resp.json()) as { sha?: string };
      return data.sha ?? null;
    } catch (e) {
      if (e instanceof RepoStoreNotFound) {
        return null;
      }
      throw e;
    }
  }

  async createRun(runId: string): Promise<void> {
    const sha = await this.getDefaultBranchSha();
    const url = `${BASE_URL}/repos/${this.owner}/${this.repo}/git/refs`;
    const payload = { ref: `refs/heads/run/${runId}`, sha };
    try {
      await this.request("POST", url, payload);
    } catch (e) {
      if (e instanceof RepoStoreError && e.statusCode === 422) {
        return;
      }
      throw e;
    }
  }

  async readFile(runId: string, filename: string): Promise<string> {
    const url = `${BASE_URL}/repos/${this.owner}/${this.repo}/contents/${filename}?ref=${encodeURIComponent(`run/${runId}`)}`;
    try {
      const resp = await this.request("GET", url);
      const data = (await resp.json()) as { content?: string };
      const encoded = data.content ?? "";
      return decodeContent(encoded);
    } catch (e) {
      if (e instanceof RepoStoreNotFound) {
        return "";
      }
      throw e;
    }
  }

  async writeFile(runId: string, filename: string, content: string): Promise<void> {
    const url = `${BASE_URL}/repos/${this.owner}/${this.repo}/contents/${filename}`;
    const branch = `run/${runId}`;
    const encoded = encodeContent(content);
    for (let attempt = 0; attempt <= MAX_RETRIES_409; attempt++) {
      const sha = await this.getFileSha(runId, filename);
      const payload: Record<string, unknown> = {
        message: `update ${filename}`,
        content: encoded,
        branch,
      };
      if (sha !== null) {
        payload.sha = sha;
      }
      try {
        await this.request("PUT", url, payload);
        return;
      } catch (e) {
        if (e instanceof RepoStoreError && (e.statusCode === 409 || e.statusCode === 422) && attempt < MAX_RETRIES_409) {
          await delay(Math.ceil(computeDelay(attempt) * 1000));
          continue;
        }
        throw e;
      }
    }
  }

  async deleteFile(runId: string, filename: string): Promise<void> {
    const sha = await this.getFileSha(runId, filename);
    if (sha === null) {
      return;
    }
    const url = `${BASE_URL}/repos/${this.owner}/${this.repo}/contents/${filename}`;
    const payload = {
      message: `delete ${filename}`,
      sha,
      branch: `run/${runId}`,
    };
    try {
      await this.request("DELETE", url, payload);
    } catch (e) {
      if (e instanceof RepoStoreNotFound) {
        return;
      }
      throw e;
    }
  }

  async listRunFiles(runId: string): Promise<string[]> {
    const url = `${BASE_URL}/repos/${this.owner}/${this.repo}/contents/?ref=${encodeURIComponent(`run/${runId}`)}`;
    try {
      const resp = await this.request("GET", url);
      const items = (await resp.json()) as { name?: string }[];
      if (!Array.isArray(items)) return [];
      return items.filter((x) => typeof x === "object" && x && "name" in x).map((x) => x.name as string);
    } catch (e) {
      if (e instanceof RepoStoreNotFound) {
        return [];
      }
      throw e;
    }
  }

  private async getBranchSha(runId: string): Promise<string> {
    const url = `${BASE_URL}/repos/${this.owner}/${this.repo}/git/ref/heads/run/${runId}`;
    const resp = await this.request("GET", url);
    const data = (await resp.json()) as { object?: { sha?: string } };
    return data.object?.sha ?? "";
  }

  private async createBlob(content: string): Promise<string> {
    const url = `${BASE_URL}/repos/${this.owner}/${this.repo}/git/blobs`;
    const resp = await this.request("POST", url, { content, encoding: "utf-8" });
    const data = (await resp.json()) as { sha?: string };
    return data.sha ?? "";
  }

  private async createTree(fileBlobs: Record<string, string>): Promise<string> {
    const tree = Object.entries(fileBlobs).map(([path, blobSha]) => ({
      path,
      mode: "100644",
      type: "blob",
      sha: blobSha,
    }));
    const url = `${BASE_URL}/repos/${this.owner}/${this.repo}/git/trees`;
    const resp = await this.request("POST", url, { tree });
    const data = (await resp.json()) as { sha?: string };
    return data.sha ?? "";
  }

  private async createCommit(message: string, treeSha: string, parentSha: string): Promise<string> {
    const url = `${BASE_URL}/repos/${this.owner}/${this.repo}/git/commits`;
    const resp = await this.request("POST", url, { message, tree: treeSha, parents: [parentSha] });
    const data = (await resp.json()) as { sha?: string };
    return data.sha ?? "";
  }

  private async updateBranchRef(runId: string, commitSha: string): Promise<void> {
    const url = `${BASE_URL}/repos/${this.owner}/${this.repo}/git/refs/heads/run/${runId}`;
    await this.request("PATCH", url, { sha: commitSha });
  }

  async listRunBranches(): Promise<string[]> {
    const url = `${BASE_URL}/repos/${this.owner}/${this.repo}/git/refs/heads/run/`;
    try {
      const resp = await this.request("GET", url);
      const items = (await resp.json()) as { ref?: string }[];
      if (!Array.isArray(items)) return [];
      return items
        .filter((x) => typeof x === "object" && x && "ref" in x)
        .map((x) => (x.ref as string).replace(/^refs\/heads\//, ""));
    } catch (e) {
      if (e instanceof RepoStoreNotFound) {
        return [];
      }
      throw e;
    }
  }

  async deleteRunBranch(runId: string): Promise<void> {
    const url = `${BASE_URL}/repos/${this.owner}/${this.repo}/git/refs/heads/run/${runId}`;
    try {
      await this.request("DELETE", url);
    } catch (e) {
      if (e instanceof RepoStoreNotFound) {
        return;
      }
      throw e;
    }
  }

  async batchWriteFiles(runId: string, files: Record<string, string>, message = "batch update"): Promise<void> {
    for (let attempt = 0; attempt <= MAX_RETRIES_409; attempt++) {
      const parentSha = await this.getBranchSha(runId);
      const fileBlobs: Record<string, string> = {};
      for (const [p, content] of Object.entries(files)) {
        fileBlobs[p] = await this.createBlob(content);
      }
      const treeSha = await this.createTree(fileBlobs);
      const commitSha = await this.createCommit(message, treeSha, parentSha);
      try {
        await this.updateBranchRef(runId, commitSha);
        return;
      } catch (e) {
        if (e instanceof RepoStoreError && e.statusCode === 422 && attempt < MAX_RETRIES_409) {
          await delay(Math.ceil(computeDelay(attempt) * 1000));
          continue;
        }
        throw e;
      }
    }
  }
}
